lizrd/core/llm.py

```diff
@@ -128,102 +130,78 @@ def LowRank(dinput, doutput, dlowrank):
+
+        a = torch.einsum("... l h d, ... L h d -> ... h l L", query, key)
+        a = a * (1 / dhead**0.5)
+        if causal:
+            a.masked_fill_(
+                torch.tril(torch.ones_like(a)) == 0, float("-inf")
+            )  # mask out future tokens
+        a = torch.softmax(a, dim=-1)
+        prefinal = torch.einsum("... h l L, ... L h d -> ... l h d", a, value)
+        prefinal = prefinal.transpose(1, 2)
+
+    return prefinal
+
+
 @ash.check("... d -> ... d")
-class Attention(nn.Module):
-    def __init__(self, dmodel, heads, dhead=None):
+class Attention(LoggingLayer):
+    def __init__(self, dmodel, heads, causal, dhead=None, flash=False):
         super(Attention, self).__init__()
         if dhead is None:
             assert dmodel % heads == 0
             dhead = dmodel // heads
 
         self.heads = heads
         self.dhead = dhead
-        self.dmodel = dmodel
-
-        key_query_value_gen = lambda: misc.EinMix(
-            "... dmodel -> ... heads dhead",
-            weight_shape="dmodel heads dhead",
-            bias_shape="heads dhead",
-            dmodel=dmodel,
-            heads=heads,
-            dhead=dhead,
-        )
+        self.causal = causal
+        self.flash = flash
 
-        self.Q = key_query_value_gen()
-        self.K = key_query_value_gen()
-        self.V = key_query_value_gen()
-
-        combine_gen = lambda: misc.EinMix(
-            "... heads dhead -> ... dmodel",
-            weight_shape="heads dhead dmodel",
-            bias_shape="dmodel",
-            dmodel=dmodel,
-            heads=heads,
-            dhead=dhead,
-        )
-        self.D = combine_gen()
+        self.input_projection = nn.Linear(dmodel, 3 * heads * dhead)
+        self.output_projection = nn.Linear(heads * dhead, dmodel)
 
     def forward(self, x):
-        q = self.Q(x)
-        k = self.K(x)
-        v = self.V(x)
-
-        a = torch.einsum("... l h d, ... L h d -> ... h l L", q, k)
-        a = a * (1 / self.dhead**0.5)
-        a = torch.softmax(a, dim=-1)
-        prefinal = torch.einsum("... h l L, ... L h d -> ... l h d", a, v)
-        output = self.D(prefinal)
-        return output
-
-
-@ash.check("... d -> ... d")
-class CausalAttention(nn.Module):
-    def __init__(self, dmodel, heads, dhead=None):
-        super(CausalAttention, self).__init__()
-        if dhead is None:
-            assert dmodel % heads == 0
-            dhead = dmodel // heads
-
-        self.heads = heads
-        self.dhead = dhead
-        self.dmodel = dmodel
-
-        key_query_value_gen = lambda: misc.EinMix(
-            "... dmodel -> ... heads dhead",
-            weight_shape="dmodel heads dhead",
-            bias_shape="heads dhead",
-            dmodel=dmodel,
-            heads=heads,
-            dhead=dhead,
+        projected = self.input_projection(x)
+
+        batch, seq_len = x.shape[:-1]
+        projected = projected.view(batch, self.heads, seq_len, 3 * self.dhead)
```

Comment:
this is pretty reckless imo. As the `projected` is originally `[batch_size, seqlen, 3 * dhead * n_heads]`, you are trying to implicitly reshape and transpose by `.view`. Your implementation almost surely will mix projections from different tokens in a sequence.
```suggestion
        projected = projected.view(batch, seq_len,  self.heads,  3 * self.dhead).transpose(1,2)
```

Reply:
You are right, fixed!